src/game/market.rs

```diff
@@ -160,41 +185,102 @@ pub fn change_order_price(order_id: &str, new_price: f64) -> ReturnCode {
 
 pub fn create_order(
     order_type: OrderType,
-    resource_type: ResourceType,
+    resource_type: MarketResourceType,
     price: f64,
     total_amount: u32,
-    room: &Room,
+    room: Option<RoomName>,
 ) -> ReturnCode {
-    js_unwrap! {
-        Game.market.createOrder(__order_type_num_to_str(@{order_type as u32}),
-                                __resource_type_num_to_str(@{resource_type as u32}),
-                                @{price},
-                                @{total_amount},
-                                @{room.name()})
+    let resource_num = match resource_type {
+        MarketResourceType::Resource(ty) => ty as u32,
+        MarketResourceType::IntershardResource(ty) => ty as u32,
+    };
+    match room {
+        Some(room_name) => {
+            js_unwrap! {
+                Game.market.createOrder({
+                    type: __order_type_num_to_str(@{order_type as u32}),
+                    resourceType: __resource_type_num_to_str(@{resource_num}),
+                    price: @{price},
+                    totalAmount: @{total_amount},
+                    roomName: @{room_name.to_string()}
+                })
+            }
+        }
+        None => {
+            js_unwrap! {
+                Game.market.createOrder({
+                    type: __order_type_num_to_str(@{order_type as u32}),
+                    resourceType: __resource_type_num_to_str(@{resource_num}),
+                    price: @{price},
+                    totalAmount: @{total_amount}
+                })
+            }
+        }
     }
 }
 
-pub fn deal(order_id: &str, amount: u32, target_room: &Room) -> ReturnCode {
-    js_unwrap! {Game.market.deal(@{order_id}, @{amount}, @{target_room.name()})}
+/// Execute a market trade
+///
+/// `target_room` is your owned room whose terminal will send or receive
+/// resources in this transaction, or `None` if this is an order for an
+/// intershard resource type
+pub fn deal(order_id: &str, amount: u32, target_room: Option<RoomName>) -> ReturnCode {
```

Comment:
Actually, should this (and the others like it) be `Option<&RoomName>` instead?

Reply:
Since `RoomName` is Copy, this should be fine. It's a very small struct (not bigger than `u32`), and practically free to copy around.